### PR TITLE
Optimise contracts

### PR DESCRIPTION
**Pull request by Holymir**
Validator issue - "Optimise Smart Contracts" #203

- Added burnWithPermit functionality
- Add the Router contract address as part of the signature

### CODE REVIEW

File: contracts/Router/Router.sol [hunk 1]
```diff
@@ -141,6 +149,38 @@ contract Router is Governance {
...
     }
 
+    /**
+     * @notice Approves and Burns the provided `amount` of wrapped tokens and emits Burn event
+     * @param wrappedAsset The corresponding wrapped asset contract
+     * @param receiver The Hedera account to receive the wrapped tokens
+     * @param amount The amount of wrapped tokens to be bridged
+     * @param deadline Timestamp of the deadline
+     */
+    function burnWhithPermit(
```
Comment (failfmi) [inline contracts/Router/Router.sol:159]: `burnWithPermit`

File: README.md [hunk 2]
```diff
@@ -73,6 +73,12 @@ npx hardhat add-pair --router /The address of the deployed router/ --native /hed
 npx hardhat remove-pair --router /The address of the deployed router/ --native /hedera token id/ --wrapped /The address of the deployed token/ --network /The name of the network/
 ```
 
+#### Remove token pair
+
+```
+npx hardhat remove-pair --router /The address of the deployed router/ --native /hedera token id/ --wrapped /The address of the deployed token/ --network /The name of the network/
+```
+
```
Comment (failfmi) [inline README.md:81]: You have this twice

**Review by failfmi — DISMISSED** [review]

File: contracts/ERC20Permit.sol [hunk 3]
```diff
@@ -0,0 +1,195 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.7.0;
+
+import "@openzeppelin/contracts/token/ERC20/ERC20Pausable.sol";
+import "@openzeppelin/contracts/utils/Counters.sol";
+
+import {IERC2612Permit} from "./Interfaces/IERC2612Permit.sol";
+
+// import "@nomiclabs/buidler/console.sol";
```
Comment (Daniel-K-Ivanov) [inline contracts/ERC20Permit.sol:9]: nit: we do not need this

File: contracts/Router/Router.sol [hunk 4]
```diff
@@ -1,9 +1,11 @@
-pragma solidity ^0.6.0;
+// SPDX-License-Identifier: MIT
```
Comment (Daniel-K-Ivanov) [inline contracts/Router/Router.sol:1]: nit: Lets move the `Router` contract outside of the folder. It seems inconsistent having it in a separate folder from the other contracts

File: contracts/WrappedToken.sol [hunk 5]
```diff
@@ -1,13 +1,15 @@
...
-import "@openzeppelin/contracts/token/ERC20/ERC20Pausable.sol";
+import "./ERC20Permit.sol";
 
 /**
  *  @author LimeChain Dev team
  *  @title ERC20 WrappedToken contract
  */
-contract WrappedToken is ERC20Pausable, Ownable {
+contract WrappedToken is ERC20Permit, Ownable {
+    using SafeMath for uint256;
```
Comment (Daniel-K-Ivanov) [inline contracts/WrappedToken.sol:12]: Q: Is this necessary?

File: hardhat.config.js [hunk 6]
```diff
@@ -96,4 +97,9 @@ module.exports = {
     mocha: {
         timeout: 20000,
     },
+    gasReporter: {
+        currency: "USD",
+        gasPrice: 256,
+        enabled: true,
+    },
```
Comment (Daniel-K-Ivanov) [inline hardhat.config.js:104]: Q: Why do we need this?

File: test/router.js [hunk 7]
```diff
@@ -380,8 +380,14 @@ describe("Router", function () {
 
         it("Should execute mint", async () => {
```
Comment (Daniel-K-Ivanov) [inline test/router.js:381]: I think that we should add test, checking that if we create a signature that does not sign the router contract -> the recovery will fail and tx will get reverted. This way we will be sure that if we upgrade the contract, there won't be replay attacks on the newly deployed one

File: contracts/WrappedToken.sol (hunk 5 above)
Comment (Holymir) [reply contracts/WrappedToken.sol:12]: yes, we need for solidity < 0.8

File: test/router.js [hunk 8]
```diff
@@ -32,6 +32,14 @@ describe("Router", function () {
...
             mockAsset,
             mockNative,
+            d,
+            e,
+            f,
+            g,
+            h,
+            i,
+            j,
+            k,
```
Comment (Daniel-K-Ivanov) [inline test/router.js:42]: I think that this is unnecessary change

File: test/router.js [hunk 9]
```diff
@@ -371,6 +379,12 @@ describe("Router", function () {
             await routerInstance.updateMember(aliceMember.address, true);
             await routerInstance.updateMember(bobMember.address, true);
             await routerInstance.updateMember(carlMember.address, true);
+            // await routerInstance.updateMember(d.address, true);
+            // await routerInstance.updateMember(e.address, true);
+            // await routerInstance.updateMember(f.address, true);
+            // await routerInstance.updateMember(g.address, true);
+            // await routerInstance.updateMember(h.address, true);
+            // await routerInstance.updateMember(i.address, true);
```
Comment (Daniel-K-Ivanov) [inline test/router.js:387]: unnecessary

File: test/router.js [hunk 10]
```diff
@@ -395,6 +409,12 @@ describe("Router", function () {
             const aliceSignature = await aliceMember.signMessage(hashData);
             const bobSignature = await bobMember.signMessage(hashData);
             const carlSignature = await carlMember.signMessage(hashData);
+            const dSignature = await d.signMessage(hashData);
+            const eSignature = await e.signMessage(hashData);
+            const fSignature = await f.signMessage(hashData);
+            const gSignature = await g.signMessage(hashData);
+            const hSignature = await h.signMessage(hashData);
+            const iSignature = await i.signMessage(hashData);
```
Comment (Daniel-K-Ivanov) [inline test/router.js:417]: unnecessary

File: test/router.js [hunk 11]
```diff
@@ -403,15 +423,25 @@ describe("Router", function () {
...
+                    [
+                        aliceSignature,
+                        bobSignature,
+                        carlSignature,
+                        // dSignature,
+                        // eSignature,
+                        // fSignature,
+                        // gSignature,
+                        // hSignature,
+                        // iSignature,
```
Comment (Daniel-K-Ivanov) [inline test/router.js:435]: unnecessary

File: contracts/Router.sol [hunk 12]
```diff
@@ -272,17 +263,20 @@ contract Router is Governance {
...
         bytes[] memory signatures
     ) private {
-        Transaction storage transaction = mintTransfers[transactionId];
+        uint256 signersCount = signatures.length;
+        address[] memory signers = new address[](signersCount);
 
-        for (uint256 i = 0; i < signatures.length; i++) {
+        for (uint256 i = 0; i < signersCount; i++) {
             address signer = ECDSA.recover(ethHash, signatures[i]);
             require(isMember(signer), "Router: invalid signer");
```
Comment (Daniel-K-Ivanov) [inline contracts/Router.sol:271]: nit: Rename the message to "Router: invalid signer/signature"

File: contracts/Router.sol [hunk 13]
```diff
@@ -272,17 +263,20 @@ contract Router is Governance {
...
-            require(
-                !transaction.hasSigned[signer],
-                "Router: signature already set"
-            );
-            transaction.hasSigned[signer] = true;
+
+            for (uint256 j = 0; j <= i; j++) {
```
Comment (Daniel-K-Ivanov) [inline contracts/Router.sol:273]: I think that we can save gas if its `j < i` and the memory `signers` length is `signersCount - 1`

**Review by Daniel-K-Ivanov — APPROVED** [review]
